lib/Canonicalize/SimplifySwitchPass.cpp

```diff
@@ -0,0 +1,282 @@
+      revng_log(Log,
+                "Have not found case/BB for "
+                  << ConstantForOld->getUniqueInteger().getZExtValue());
+      return false;
+    }
+    NewLabels[cast<ConstantInt>(ConstantForTheValue)] = BlockForValue;
+  }
+
+  // We have not found all the cases for the new switch.
+  if (Switch->getNumCases() != NewLabels.size())
+    return false;
+
+  // Recreate the new switch.
+  auto *NewSwitch = SwitchInst::Create(StartNode->Value,
+                                       Switch->getDefaultDest(),
+                                       NewLabels.size(),
+                                       Switch);
+
+  for (auto &NewLabel : NewLabels)
+    NewSwitch->addCase(NewLabel.first, NewLabel.second);
+
+  Switch->replaceAllUsesWith(NewSwitch);
+
+  return true;
+}
+
+static bool runSimplifySwitch(Function &F,
+                              LazyValueInfo &LVI,
+                              DominatorTree &DT,
+                              RawBinaryMemoryOracle &MO) {
+  bool Result = false;
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : make_early_inc_range(BB)) {
+      SwitchInst *Switch = dyn_cast<SwitchInst>(&I);
+      if (not Switch)
+        continue;
+
+      if (handleSwitch(Switch, LVI, DT, MO)) {
+        Switch->eraseFromParent();
+        Result |= true;
+      }
+    }
+  }
+
+  return Result;
+}
+
+struct SimplifySwitchPass : public FunctionPass {
+public:
+  static char ID;
+
+  SimplifySwitchPass();
+
+  bool runOnFunction(Function &F) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override;
+};
+
+SimplifySwitchPass::SimplifySwitchPass() : FunctionPass(ID) {
+}
+
+bool SimplifySwitchPass::runOnFunction(Function &F) {
+  bool Changed = false;
+
+  auto &DT = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+  auto &LVI = getAnalysis<LazyValueInfoWrapperPass>().getLVI();
+
+  RawBinaryView &BinaryView = getAnalysis<LoadBinaryWrapperPass>().get();
+  const model::Binary
+    &Model = *getAnalysis<LoadModelWrapperPass>().get().getReadOnlyModel();
+
+  RawBinaryMemoryOracle MO(BinaryView, Model.Architecture());
+  Changed |= runSimplifySwitch(F, LVI, DT, MO);
+
+  return Changed;
+}
+
+void SimplifySwitchPass::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.addRequired<DominatorTreeWrapperPass>();
+  AU.addRequired<LazyValueInfoWrapperPass>();
+  AU.addRequired<LoadBinaryWrapperPass>();
+  AU.addRequired<LoadModelWrapperPass>();
+}
+
+char SimplifySwitchPass::ID = 0;
+static constexpr const char *Flag = "simplify-switch";
+using Register = RegisterPass<SimplifySwitchPass>;
+static Register X(Flag, "Pass that simplifies switch statement.", false, false);
+
+namespace revng::pipes {
+
+class SimplifySwitch {
+public:
+  static constexpr auto Name = "simplify-switch";
+
+  std::array<pipeline::ContractGroup, 1> getContract() const {
+    pipeline::Contract BinaryPart(kinds::Binary, 0, kinds::Binary, 0);
+    pipeline::Contract FunctionsPart(kinds::Isolated, 1);
```

Comment:
You need to use the kind produced by the last pipe, i.e., `kinds::StackPointerPromoted,`.